src/app/dao/tests/JoinerDAO.spec.ts

```diff
@@ -1,17 +1,438 @@
+        const candidate: MinimalUser = { id: candidateUser.uid, name: 'candidate' };
+        // Add ourself to candidate list to try to have more privileges
+        await joinerDAO.addCandidate(partId, candidate);
+
+        const updates: Partial<Joiner>[] = [
+            { chosenOpponent: candidate },
+            { partStatus: PartStatus.CONFIG_PROPOSED.value },
+            { firstPlayer: FirstPlayer.CHOSEN_PLAYER.value },
+            { partType: PartType.BLITZ.value },
+            { maximalMoveDuration: 73 },
+            { totalPartDuration: 1001001 },
+        ];
+        for (const update of updates) {
+            // When modifying a field
+            const result: Promise<void> = joinerDAO.update(partId, update);
+            // Then it should fail
+            await expectFirebasePermissionDenied(result);
+        }
+    });
+    describe('chosen opponent', () => {
+        let partId: string;
+        beforeEach(async() => {
+            // Given a part with a joiner where the current user is set as chosen opponent
+
+            // We have to follow a realistic workflow to achieve that
+            // The part is first created
+            const creatorUser: FireAuth.User = await createConnectedGoogleUser('foo@bar.com', 'creator');
+            const creator: MinimalUser = { id: creatorUser.uid, name: 'creator' };
+            partId = await partDAO.create(PartMocks.INITIAL);
+            await joinerDAO.set(partId, { ...JoinerMocks.INITIAL, creator });
+            await signOut();
+
+            // A candidate add itself to the candidates list
+            const candidateUser: FireAuth.User = await createConnectedGoogleUser('bar@bar.com', 'candidate');
+            const candidate: MinimalUser = { id: candidateUser.uid, name: 'candidate' };
+            await expectAsync(joinerDAO.addCandidate(partId, candidate)).toBeResolvedTo();
+            await signOut();
+
+            // The creator then selects candidates as the chosen opponent and proposes the config
+            await reconnectUser('foo@bar.com');
+            const update: Partial<Joiner> = { chosenOpponent: candidate, partStatus: PartStatus.CONFIG_PROPOSED.value };
+            await expectAsync(joinerDAO.update(partId, update)).toBeResolvedTo();
+            await signOut();
+
+            // We will finally act as the chosen opponent
+            await reconnectUser('bar@bar.com');
+        });
+        it('should allow chosen opponent to change status from PROPOSED to STARTED', async() => {
+            // When the candidate accepts the config by setting partStatus
+            const result: Promise<void> = joinerDAO.update(partId, { partStatus: PartStatus.PART_STARTED.value });
+
+            // Then it should succeed
+            await expectAsync(result).toBeResolvedTo();
+        });
+        it('should forbid chosen opponent to change status to any other value than STARTED', async() => {
```

Comment:
Donc si le createur timeout et perd, qui set la partie à FINISHED :o ?

Reply:
good point! Une idée de comment résoudre ça pour éviter de laisser un joueur changer la partie à FINISHED sans qu'elle soit vraiment finie ? Il me semble que c'est jeu-dépendant, donc pour vraiment savoir qui peut mettre la partie à FINISHED, il faut prendre en compte les règles; et c'est out-of-scope de ce ticket alors -> sécu v2 avec le backend ! Si tu agree, le changement qu'il me reste à faire c'est s'assurer qu'on puisse terminer une partie comme tu as remarqué :D

Reply:
Bah, tout joueur peux gagner, et c'est en gagnant qu'on envoi les champs lié à la fin de partie dans la Part.
Mais en fait pour l'instant personne ne fait jamais passer le joiner à FINISHED haha :D 
EN vérité, un joiner c'est une pré-partie, il passe de "en cours de config" à "partie commencée" et c'est la partie qui elle passe de "en attente de la config" à "en cours" puis à "finie". On pourrais pour un futur ticket dire que au même moment que la partie change le joiner aussi (et donc ce champs seraient un peu dupliqués) ou que le rôle du joiner s'arrête dès que la partie commence (et auquel cas FINISHED disparaît, vu que inutilisé)
Qu'est-ce-t'en-pense-tu ?

Reply:
Hmm et quid de supprimer le joiner a un moment ? Ce qui compte, c'est de préserver la partie et éventuellement le chat. Le joiner il pourrait très bien être supprimé à la fin de la partie, voir même quand la partie commence. En gros, la ConfigRoom existe tant qu'on configure, une fois que la partie est lancée c'est une Part. On pourrait même avoir une collection FinishedPart: quand une partie est finie, on copie la partie + le chat dans une FinishedPart dans un but d'archivage, et on supprime partie + chat aussi. Y'a discussion intéressante sur le sujet à avoir, peut-être dans un autre ticket du coup ?

Reply:
Supprimer la ConfigRoom après le début de la partie pourquoi pas, merger les deux peut être pas, c'est p-e mieux de pas avoir des "document mêle tout", ils y a déjà pas mal de champs dans Part.

J'avais pensé que les finished part soient bien stockée, mais p-e dans une table relationnelle avec d'autres buts (stats, history, level calculation).

Mais oui, j'suis partagé entre "l'history c'est cool" et "ne pas faire fondre toute la norvège, c'est un site à vivre pas à disque-duriser" :D

Reply:
ça vaudrait la peine de bouger des infos de la configRoom à la part ceci dit, pour diminuer le nombre de requêtes (genre si tu as une part et que tu veux savoir qui est créateur, tu dois aller voir la configRoom, alors que ça pourrait bien être stocké dans la part). Si on a toutes les infos essentielles, on peut safement supprimer la configRoom lorsque la partie est démarrée, et garder les parties en mémoire. Tout ce qui est stats supplémentaires me semble calculable après coup à partir de l'historique de la partie non ?

On tickette ceci ? (suppression de configroom à la création)

Reply:
Tu peux ticketter le transfer des données dans la Part et la destruction de la configroom à la création de la part oui !!